plugin_tests/sources_test.py

```diff
@@ -0,0 +1,182 @@
+#!/usr/bin/env python
+# -*- coding: utf-8 -*-
+
+#############################################################################
+#  Copyright Kitware Inc.
+#
+#  Licensed under the Apache License, Version 2.0 ( the "License" );
+#  you may not use this file except in compliance with the License.
+#  You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+#  Unless required by applicable law or agreed to in writing, software
+#  distributed under the License is distributed on an "AS IS" BASIS,
+#  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+#  See the License for the specific language governing permissions and
+#  limitations under the License.
+#############################################################################
+
+import numpy
+import os
+import PIL.Image
+
+from girder import config
+from tests import base
+
+from . import common
+
+
+# boiler plate to start and stop the server
+
+os.environ['GIRDER_PORT'] = os.environ.get('GIRDER_TEST_PORT', '20200')
+config.loadConfig()  # Must reload config to pickup correct port
+
+
+def setUpModule():
+    base.enabledPlugins.append('large_image')
+    base.startServer(False)
+
+
+def tearDownModule():
+    base.stopServer()
+
+
+class LargeImageSourcesTest(common.LargeImageCommonTest):
+    def testMagnification(self):
+        file = self._uploadFile(os.path.join(
+            os.environ['LARGE_IMAGE_DATA'], 'sample_jp2k_33003_TCGA-CV-7242-'
+            '11A-01-TS1.1838afb1-9eee-4a70-9ae3-50e3ab45e242.svs'))
+        itemId = str(file['itemId'])
+        item = self.model('item').load(itemId, user=self.admin)
+        source = self.model('image_item', 'large_image').tileSource(item)
+        mag = source.getMagnification()
+        self.assertEqual(mag['magnification'], 40.0)
+        self.assertEqual(mag['mm_x'], 0.000252)
+        self.assertEqual(mag['mm_y'], 0.000252)
+        mag = source.getMagnificationForLevel()
+        self.assertEqual(mag['magnification'], 40.0)
+        self.assertEqual(mag['mm_x'], 0.000252)
+        self.assertEqual(mag['mm_y'], 0.000252)
+        self.assertEqual(mag['level'], 7)
+        mag = source.getMagnificationForLevel(0)
+        self.assertEqual(mag['magnification'], 0.3125)
+        self.assertEqual(mag['mm_x'], 0.032256)
+        self.assertEqual(mag['mm_y'], 0.032256)
+        self.assertEqual(mag['level'], 0)
+        self.assertEqual(source.getLevelForMagnification(), 7)
+        self.assertEqual(source.getLevelForMagnification(exact=True), None)
```

Comment:
I can change the `exact=True` case.  I guess my thinking was that if you asked for an exact magnification but then didn't specify anything, that would be an ill-defined request (hence the None return), but if you didn't care if it were exact, then in some sense, any level would do (so return the default, which is the highest level).

levels runs from from 0 (the single tile instance) to <number of levels> - 1 (the maximum resolution).